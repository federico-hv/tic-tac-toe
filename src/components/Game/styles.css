
@font-face {
  font-family: Comic Sans MS3;
  src: url(../../assets/comicsansms3.ttf);
}

body {
	font-family: Comic Sans MS3;
	margin: 0;
}

.game-container {
	height: 100vh;
	display: flex;
	justify-content: center;
	align-items: center;
}

.gc-1 {
	background-color: black;
}

.gc-2 {
	background-color: white;
}

.board-container {
	display: flex;
	justify-content: center;
	flex-wrap: wrap;
	width: 100vw;
	height: 80vh;
	position: absolute;
}

.piece-selector {
	display: flex;
	flex-direction: column;
}

.piece-selector-title {
	display: flex;
}

.toe {
	margin-left: 20px;
}

.piece-selector-els {
	display: flex;
	flex-direction: column;
	justify-content: center;
	align-items: center;
	font-size: 40px;
	color: #16ECC5;
	margin-top: 20px;
}

.btns-container {
	display: flex;
	justify-content: space-around;
	margin-top: 40px;
}

#btnX, #btnO {
	cursor: pointer;
	display: flex;
	justify-content: center;
	align-items: center;
	width: 100px;
	height: 100px;
	font-family: Comic Sans MS3;
	font-size: 3.5em;
	padding-bottom: 0.8em;
	outline:none;
	border: none;
	border-radius: 5px;
	box-shadow: 0px 17px 10px -10px rgba(0,0,0,.3);
	transition: all 300ms;
}

#btnX:hover, #btnO:hover {
	background: rgb(88,154,204);
	transform: translate(0px, -2px) scale(1.05);
	color: white;
	box-shadow: 
}

#btnX:active, #btnO:active {
	background: black;
	outline:none;
	border: none;
}

#btnX {
	color: #16ECC5;
	background-color: #EB7272;
}

#btnO {
	color: #EB7272;
	background-color: #16ECC5;
}
