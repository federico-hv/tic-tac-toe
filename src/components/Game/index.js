import React, { Component } from 'react';
import Board from '../Board';
import Piece from '../Piece';
import './styles.css';


class Game extends Component {

  state = {
    data: (new Array(3)).fill().map(row => new Array(3).fill('')),
    starterPiece: null,
    player: 0,
    gameEnded: false,
    result: ''
  };

  selectPiece = (p) => this.setState({starterPiece: p})

  /**
   * Adds a value to the matrix using the coordinates
   * and the player id 0 or 1
   *
   * @param {array} pos array with coordinates of recent play
   */
  onPlayerMove = (pos) => {
    const { data, player, starterPiece } = this.state;
    data[pos[0]][pos[1]] = player === starterPiece ? 'X' : 'O';
    
    this.setState({ data }, this.checkGameResult);
  }

  /**
   * Receives the data array and returns the final result
   * of the game.
   * 
   * 0 if game hasn't finished yet
   * 1 current player wins
   * 2 for tie
   *
   * @return  {number}  current status of the game
   */
  getGameResult = (data) => {
    let result = 0, missing = false, aux1 = [], aux2 = [], aux3 =[];

    data.every(row => {
      if (row.every(v => v === row[0] && v !== '')){
        result = 1;
        return false
      }

      if (row.some(v => v === '')) {
        missing = true;
      }

      aux1.push(row[0]);
      aux2.push(row[1]);
      aux3.push(row[2]);
      return true;
    })

    if (result === 1) return result;
    
    if (aux1.every( v => v === aux1[0] && v !== '') || 
        aux2.every( v => v === aux2[0] && v !== '') || 
        aux3.every( v => v === aux3[0] && v !== ''))
      return 1;

    if ((data[0][0] === data[1][1] && data[1][1] === data[2][2] && data[0][0] !== '') || 
        data[0][2] ===Â data[1][1] &&  data[1][1] === data [2][0] && data[0][2] !== '') 
        return 1;

    if (!missing) return 2;

    return 0;
  }

  /**
   * Checks and updates the state of the game
   * after every move
   */
  checkGameResult = () => {

    const { data } = this.state;

    const result = this.getGameResult(data);
    const newPlayer = this.state.player === 0 ? 1 : 0;
    
    switch (result) {
      case 0:
        return this.setState({ player: newPlayer });
      case 1:
        return this.setState({ gameEnded: true, result: `Player ${this.state.player+1} wins`});
      case 2:
        return this.setState({ gameEnded: true, result: `There's a tie`});
      default:
        return this.setState({ player: newPlayer });
    }
  }

  render() {
    return (
      <div className="game-container">
        {
          this.state.gameEnded ? 
            (
              <div className="game-result">{this.state.result}</div>
            )
          : (
            <div className="board">
              <h1 className="player-turn">{this.state.player === 0 ? 'First Player' : 'Second Player'}</h1>
              {this.state.starterPiece !== null ? <Board onMove={this.onPlayerMove} data={this.state.data} /> : 
                (
                  <div className="piece-selector">
                    Select 
                    <button id="btnX" onClick={()=>this.selectPiece(0)}>x</button>
                    <button id="btnO" onClick={()=>this.selectPiece(1)}>o</button>
                  </div>
                )
              }
            </div>
          )
        }
        
      </div>
    );
  }
}


export default Game;

export {
  Board,
  Piece
};
