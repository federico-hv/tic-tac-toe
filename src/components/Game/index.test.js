import React from 'react';
import { mount, shallow } from 'enzyme';
import Game, { Piece, Board } from './index';
import toJson from 'enzyme-to-json';


describe('<Game />', ()=>{

	it('Should render a game container if game started', () => {
		const wrapper = shallow(<Game />);
		expect(wrapper.find('.board').length).toEqual(1);
		wrapper.unmount();
	});
	
	it('Should update game data properly onPlayerMove', () => {	
		const app = shallow(<Game />);
		const data = app.state().data;
	
		app.setState({ player: 0, starterPiece: 0 })
		expect(data).toEqual([ [ '', '', '' ], [ '', '', '' ], [ '', '', '' ] ])
		app.instance().onPlayerMove([2,2]);
		expect(data).toEqual([ [ '', '', '' ], [ '', '', '' ], [ '', '', 'X' ] ])
		app.setState({ player: 1 })
		app.instance().onPlayerMove([1,1]);
		expect(data).toEqual([ [ '', '', '' ], [ '', 'O', '' ], [ '', '', 'X' ] ])
	})

	it('Should return the game result', () => {
		const app = shallow(<Game />);
		let data = app.state().data;
		
		//default data means game is still on
		expect(app.instance().getGameResult(data)).toEqual(0);
		
		// an array with a winning result should return 1
		data = [['X', '', 'O'],['O','X',''],['X','O','X']]
		app.setState({ data });
		expect(app.instance().getGameResult(data)).toEqual(1);

		// an array with a tie result should return 2
		data = [['X', 'O', 'X'],['O','O','X'],['X','X','O']]
		app.setState({ data });
		expect(app.instance().getGameResult(data)).toEqual(2);
	})


	it('Should render the final result after checkGameResult when game ends', () => {
		const app = shallow(<Game />);
		const data = [['X', '', 'O'],['X','O',''],['X','O','X']];
		
		app.setState({ data });
		app.instance().getGameResult = jest.fn().mockReturnValueOnce(1);
		app.instance().checkGameResult();
		expect(app.find('.board').length).toEqual(0);
		expect(app.find('.game-result').length).toEqual(1);
	})

	
	it('Should render the turn of the corresponding player on every turn', () => {
		const app = shallow(<Game />);
		const data = [['X', '', 'O'],['X','O',''],['X','O','X']];
		
		expect(app.find('.player-turn').length).toEqual(1);
		expect(app.find('.player-turn').text()).toEqual('First Player');

		app.instance().getGameResult = jest.fn().mockReturnValueOnce(0);
		app.instance().checkGameResult();
		
		expect(app.find('.player-turn').text()).toEqual('Second Player');
	})

	it('Should render piece selector before game starts and board after', () => {	
		const app = mount(<Game />);

		expect(app.find('.piece-selector').length).toBe(1);
		
		const button = app.find('#btnX');
		button.simulate('click');
		
		expect(app.find('.piece-selector').length).toBe(0);

	})

	it('Matches snapshot', () => {
		const tree = mount(<Game />);
		expect(toJson(tree)).toMatchSnapshot();
		tree.unmount();
	});
	
});


describe('<Board />', () => {

	it('Should a div with class board', ()=> {
		const data = [['X', '', 'O'],['O','X',''],['X','O','X']]
		const board = shallow(<Board data={data} />);

		expect(board.find('.board').length).toEqual(1);
	})

	it('Matches snapshot', () => {
		const data = [['X', '', 'O'],['O','X',''],['X','O','X']]
		const board = shallow(<Board data={data} />);
		
		expect(toJson(board)).toMatchSnapshot();
		board.unmount();
	});
})


describe('<Piece />', () => {

	it('Should render an empty space', () => {
		const piece = shallow(<Piece symbol={{ value: null }} />)
		expect(piece.text()).toEqual('')
	});

	it('Should render an X', () => {
		const piece = shallow(<Piece symbol={'X'} />)
		expect(piece.text()).toEqual('X')
	});

	it('Matches snapshot', () => {
		const tree = mount(<Piece />);
		expect(toJson(tree)).toMatchSnapshot();
		tree.unmount();
	});
})