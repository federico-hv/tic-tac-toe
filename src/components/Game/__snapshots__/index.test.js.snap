// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Board /> Matches snapshot 1`] = `
<div
  className="board"
>
  <Piece
    key="0"
    onSelect={[Function]}
    symbol="X"
  />
  <Piece
    key="1"
    onSelect={[Function]}
    symbol=""
  />
  <Piece
    key="2"
    onSelect={[Function]}
    symbol="O"
  />
  <Piece
    key="0"
    onSelect={[Function]}
    symbol="O"
  />
  <Piece
    key="1"
    onSelect={[Function]}
    symbol="X"
  />
  <Piece
    key="2"
    onSelect={[Function]}
    symbol=""
  />
  <Piece
    key="0"
    onSelect={[Function]}
    symbol="X"
  />
  <Piece
    key="1"
    onSelect={[Function]}
    symbol="O"
  />
  <Piece
    key="2"
    onSelect={[Function]}
    symbol="X"
  />
</div>
`;

exports[`<Game /> Matches snapshot 1`] = `
<Game>
  <div>
    <div
      className="board"
    >
      <h1
        className="player-turn"
      >
        First Player
      </h1>
      <div
        className="piece-selector"
      >
        Select
        <button
          id="btnX"
          onClick={[Function]}
        >
          x
        </button>
        <button
          id="btnO"
          onClick={[Function]}
        >
          o
        </button>
      </div>
    </div>
  </div>
</Game>
`;

exports[`<Piece /> Matches snapshot 1`] = `
<Piece>
  <div
    className="piece"
    onClick={null}
  />
</Piece>
`;
